### PR TITLE
Support for --target wasm32-unknown-emscripten

### PR DESCRIPTION
**Pull request by caiiiycuk**
**Description**
I found that adding support for `wasm32-unknown-emscripten` is not so hard. Even more implementation is not depended on 3rd party libraries like winit or stdweb. It's because emscripten it self provides complete implementation for creating OpenGL 3.0 context. So crates like glow, khronos-egl works as is. Only visible differences between emscripten and native platform is that emscripten does not support pbuffer (which is not needed) and window will created automatically.

I think supporting `wasm32-unknown-emscripten` will be a great unique feature of wgpu.

**Testing**

You can test this changes by calling `./run-wasm-emscripten-example.sh`. However you need to install [emsdk](https://github.com/emscripten-core/emsdk) first, and `emcc` should be in your path.

### CODE REVIEW

**Comment by caiiiycuk** [comment]
Can I just apply `cargo fmt` to fix formatting errors?

**Comment by kvark** [comment]
> Can I just apply cargo fmt to fix formatting errors?

yes, that's what we do.
Unless the rustfmt version on CI is very different from yours

File: Cargo.toml
Comment (kvark) [inline Cargo.toml:27]: accidental change?

File: wgpu/examples/hello-triangle-em/main.rs
Comment (kvark) [inline wgpu/examples/hello-triangle-em/main.rs:1]: does this need to be a separate example? Can we instead just have `#[cfg(target_os = "emscripten")]` within the original hello-triangle?

File: wasm-resources/index.em.template.html
Comment (kvark) [inline wasm-resources/index.em.template.html:1]: can this template be reduced, or is this a standard one?

**Review by kvark — COMMENTED** [review]
Thank you for experimenting with this!
I agree it doesn't look too bad, and we can have this support.
In addition to the notes in the review, we need a CI job to test this configuration.

File: wasm-resources/index.em.template.html
Comment (caiiiycuk) [reply wasm-resources/index.em.template.html:1]: This is standard one, however I can reduce it to simple one.

**Comment by caiiiycuk** [comment]
Blocked by grovesNL/glow#195

File: wasm-resources/index.em.template.html
Comment (kvark) [reply wasm-resources/index.em.template.html:1]: would be great!

**Comment by caiiiycuk** [comment]
Done

File: .github/workflows/ci.yml
Comment (kvark) [inline .github/workflows/ci.yml:80]: sorry, I'm doing a moderate refactor in #2249
we'll need to rebase on top.

File: Cargo.toml
Comment (kvark) [inline Cargo.toml:27]: we are blocked on this currently?

File: wasm-resources/index.em.template.html
Comment (kvark) [inline wasm-resources/index.em.template.html:1]: nice!

File: Cargo.toml
Comment (caiiiycuk) [reply Cargo.toml:27]: I don't know why maintainer of glow is not responding to my PR. I think it's better to wait until he merges it...

**Comment by caiiiycuk** [comment]
Updated. Blocker is gone, can be merged now if it's ok

File: .github/workflows/ci.yml
Comment (kvark) [inline .github/workflows/ci.yml:100]: line no longer applicable

File: .github/workflows/ci.yml
Comment (kvark) [inline .github/workflows/ci.yml:179]: I think we need to do the same as regulalr webgl - require `webgl` feature to be passed in. Otherwise, this will be confusing when we get emscripten webgpu target.

File: wgpu-hal/src/gles/egl.rs
Comment (kvark) [inline wgpu-hal/src/gles/egl.rs:523]: why is this removed?

File: wgpu-hal/src/gles/egl.rs
Comment (kvark) [inline wgpu-hal/src/gles/egl.rs:641]: wouldn't it fall under this path anyway in the previous `if/else` chain?

File: wgpu-hal/src/gles/egl.rs
Comment (kvark) [inline wgpu-hal/src/gles/egl.rs:982]: no need to `cfg` this line

File: wgpu-hal/src/gles/egl.rs
Comment (kvark) [inline wgpu-hal/src/gles/egl.rs:1024]: no need for this cfg, it will go the regular `create_window_surface` way naturally, since `self.wsi.kind == WindowKind::Unknown`

File: wgpu/Cargo.toml
Comment (kvark) [inline wgpu/Cargo.toml:87]: I don't think this configuration needs to change. Instead, it should be aligned with the current webgl path.

File: wgpu/src/backend/mod.rs
Comment (kvark) [inline wgpu/src/backend/mod.rs:3]: similarly, this configuration doesn't need to change. If emscripten path requires "webgl" feature, it mostly follows the same steps.

**Review by kvark — CHANGES_REQUESTED** [review]

**Comment by caiiiycuk** [comment]
Will do new one based on "feature"

**Pull request closed without merging**
